docs/design/reflection-flow.md

```diff
@@ -0,0 +1,202 @@
+
+(When the attribute is applied to a location of type `System.String`, the assumption is that the string represents a fully qualified type name.)
+
+```csharp
+public sealed class DynamicallyAccessedMembersAttribute : Attribute
+{
+    public DynamicallyAccessedMembersAttribute(MemberKinds memberKinds)
+    {
+        MemberKinds = memberKinds;
+    }
+
+    public MemberKinds MemberKinds { get; }
+}
+
+[Flags]
+public enum MemberKinds
+{
+    DefaultConstructor = 1,
+    PublicConstructors = 3, // Maps to BindingFlags.Public
+    Constructors = 7, // Maps to BindingFlags.Public | BindingFlags.NonPublic
+    PublicProperties = 0x10,
+    Properties = 0x30,
+    PublicMethods = 0x40,
+    Methods = 0x70,
+    NestedTypes = 0x80,
+    // ...
+}
+```
+
+When a method or field is annotated with this attribute, two things will happen:
+* The method/field becomes potentially linker-friendly. Linker will ensure that the values logically written to the annotated location (i.e. passed as a parameter, returned from the method, written to the field) can be statically reasoned about, or a warning will be generated.
+* The analysis of the value read from the annotated location will have richer information available and the linker can assume that linker-unfriendly operations with an otherwise unknown value could still be safe.
+
+Example:
+
+```csharp
+class Program
+{
+    [DynamicallyAccessedMembers(MemberKinds.Methods)]
+    private static readonly Type _otherType;
+
+    static Program()
+    {
+        // _otherType is marked DynamicallyAccessedMembers - the linker
+        // ensures that the value written to this can be statically reasoned about
+        // and meets the MemberKinds restriction. So it will keep all methods on Foo.
+        _otherType = Type.GetType("Foo, FooAssembly");
+    }
+
+    static void Main()
+    {
+        // SAFE: _otherType was annotated as "some type that has all methods kept"
+        _otherType.GetMethod("RandomMethod");
+    }
+}
+```
+
+(The above pattern exists in CoreLib and is currently unfriendly.)
+
+
+TODO:  Creating a delegate to a potentially linker unfriendly method could be solvable. Reflection invoking a potentially linker unfriendly method is hard. Both out of scope?
+TODO: It might be possible to apply similar pattern to generic parameters. The DynamizallAccessedMembers could be added to the generic parameter declaration and linker could make sure that when it's "assigned to" the requirements are met.
+
+## Intrinsic recognition of reflection APIs
+
+While it would be possible to annotate reflection primitives with the proposed DynamicallyAccessedMembers attribute, linker is going to intrinsically recongnize some of the reflection primitives so that it can do a better job at preserving just the pieces that are really needed.
+
+* `Type.GetMethod`
+* `Type.GetProperty`
+* `Type.GetField`
+* `Type.GetMember`
+* `Type.GetNestedType`
+
```

Comment:
"Are going ..." reads like a question, but it is not actually question. Maybe say "The following methods are going to be special cased by linker:" in front of the list instead.